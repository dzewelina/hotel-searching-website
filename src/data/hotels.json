[
  {
    "Name": "Spring Phoenix Resort",
    "StarRating": 5,
    "Facilities": ["car park", "restaurant", "bar"],
    "Image":"https://images.pexels.com/photos/271624/pexels-photo-271624.jpeg?w=1260&h=750&auto=compress&cs=tinysrgb"
  },
  {
    "Name": "Triple Pass Hotel",
    "StarRating": 3,
    "Facilities": ["car park", "gym"],
    "Image": "https://images.pexels.com/photos/261102/pexels-photo-261102.jpeg?w=1260&h=750&auto=compress&cs=tinysrgb"
  },
  {
    "Name": "Southern Shack Hotel & Spa",
    "StarRating": 4,
    "Facilities": ["pool", "bar", "restaurant"],
    "Image": "https://images.pexels.com/photos/338504/pexels-photo-338504.jpeg?w=1260&h=750&auto=compress&cs=tinysrgb"
  },
  {
    "Name": "Oriental Cove Resort",
    "StarRating": 2,
    "Facilities": ["car park", "pool"],
    "Image": "https://images.pexels.com/photos/6534/holiday-vacation-hotel-luxury.jpg?w=1260&h=750&auto=compress&cs=tinysrgb"
  },
  {
    "Name": "Lunar Hotel",
    "StarRating": 4,
    "Facilities": ["pool", "bar", "restaurant"],
    "Image": "https://images.pexels.com/photos/53577/hotel-architectural-tourism-travel-53577.jpeg?w=1260&h=750&auto=compress&cs=tinysrgb"
  },
  {
    "Name": "Rosewood Hotel",
    "StarRating": 1,
    "Facilities": [],
    "Image": "https://images.pexels.com/photos/186181/pexels-photo-186181.jpeg?w=1260&h=750&auto=compress&cs=tinysrgb"
  }
]