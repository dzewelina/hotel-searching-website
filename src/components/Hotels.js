import React, { Component } from 'react';
import hotels from '../data/hotels.json';
import Hotel from './Hotel';
import Checkboxes from './Checkboxes';
import _ from 'underscore';
import Sorting from './Sorting';

class Hotels extends Component {
  state = {
    hotels,
    filters: [],
    sort: 0
  };

  render() {
    const { hotels, filters, sort } = this.state;
    return (
      <div className='hotelsContainer'>
        <Checkboxes handleChange={this.updateFilters} />
        <div className='sortingAndHotels'>
          <Sorting updateSort={this.updateSort} />
          <div className='hotels'>
            <ul className='list-unstyled'>
              {hotels
                .filter(hotel => !_.difference(filters, hotel.Facilities).length)
                .sort((hotelA, hotelB) => sort * (hotelA.StarRating - hotelB.StarRating))
                .map((hotel, i) => (
                  <Hotel key={i} hotel={hotel} />
                ))}
            </ul>
          </div>
        </div>
      </div>
    );
  }

  updateSort = (e) => {
    this.setState({ sort: +e.target.value });
  }

  updateFilters = (e) => {
    const checked = e.target.value;
    let { filters } = this.state;

    if (filters.includes(checked)) filters = _.reject(filters, (filter) => filter === checked);
    else filters.push(checked);

    this.setState({ filters });
  }
}

export default Hotels;
