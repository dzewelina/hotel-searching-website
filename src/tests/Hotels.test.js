import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
import renderer from 'react-test-renderer';

import React from 'react';
import Hotels from '../components/Hotels';
import hotels from '../data/hotels.json';

describe('Hotels component', () => {
  it('matches the snapshot', () => {
    const tree = renderer.create(<Hotels />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders all hotels', () => {
    const wrapper = shallow(<Hotels />);
    const numberOfHotelsData = hotels.length;
    const numberOfHotels = wrapper.find('Hotel').length;
    expect(numberOfHotels).toEqual(numberOfHotelsData);
  });
  it('updates state when checkbox clicked', () => {
    const wrapper = mount(<Hotels />);
    wrapper.find('input[value="car park"]').simulate('change', { target: { value: 'car park' } });
    expect(wrapper.state().filters).toEqual(['car park']);

    wrapper.find('input[value="pool"]').simulate('change', { target: { value: 'pool' } });
    expect(wrapper.state().filters).toEqual(['car park', 'pool']);

    wrapper.find('input[value="car park"]').simulate('change', { target: { value: 'car park' } });
    expect(wrapper.state().filters).toEqual(['pool']);
  });
  it('filters hotels by facilities', () => {
    const testingHotels = [
      {
        'Name': 'Testing Hotel',
        'StarRating': 3,
        'Facilities': ['car park', 'restaurant', 'bar'],
        'Image': 'https://image'
      },
      {
        'Name': 'Testing Hotel2',
        'StarRating': 5,
        'Facilities': ['car park', 'pool', 'gym'],
        'Image': 'https://image2'
      }
    ];
    const wrapper = shallow(<Hotels />);
    wrapper.setState({ hotels: testingHotels, filters: ['gym'] });
    const numberOfHotels = wrapper.find('Hotel').length;
    expect(numberOfHotels).toEqual(1);

    wrapper.setState({ filters: ['gym', 'bar'] });
    const numberOfHotels2 = wrapper.find('Hotel').length;
    expect(numberOfHotels2).toEqual(0);
  });
  it('updates state when sort option chosen', () => {
    const wrapper = mount(<Hotels />);
    wrapper.find('select').simulate('change', { target: { value: '-1' } });
    expect(wrapper.state().sort).toEqual(-1);

    wrapper.find('select').simulate('change', { target: { value: '1' } });
    expect(wrapper.state().sort).toEqual(1);
  });
});