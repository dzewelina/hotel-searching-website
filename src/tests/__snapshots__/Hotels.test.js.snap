// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hotels component matches the snapshot 1`] = `
<div
  className="hotelsContainer"
>
  <div
    className="filters"
  >
    <h2>
      Filters
    </h2>
    <div
      className="form-check"
    >
      <label
        className="form-check-label"
      >
        <input
          className="form-check-input"
          onChange={[Function]}
          type="checkbox"
          value="car park"
        />
        car park
      </label>
    </div>
    <div
      className="form-check"
    >
      <label
        className="form-check-label"
      >
        <input
          className="form-check-input"
          onChange={[Function]}
          type="checkbox"
          value="restaurant"
        />
        restaurant
      </label>
    </div>
    <div
      className="form-check"
    >
      <label
        className="form-check-label"
      >
        <input
          className="form-check-input"
          onChange={[Function]}
          type="checkbox"
          value="bar"
        />
        bar
      </label>
    </div>
    <div
      className="form-check"
    >
      <label
        className="form-check-label"
      >
        <input
          className="form-check-input"
          onChange={[Function]}
          type="checkbox"
          value="gym"
        />
        gym
      </label>
    </div>
    <div
      className="form-check"
    >
      <label
        className="form-check-label"
      >
        <input
          className="form-check-input"
          onChange={[Function]}
          type="checkbox"
          value="pool"
        />
        pool
      </label>
    </div>
  </div>
  <div
    className="sortingAndHotels"
  >
    <div
      className="form-group sorting"
    >
      <label>
        Sort by: 
        <b>
          Star Rating
        </b>
      </label>
      <select
        className="form-control col-md-2"
        onChange={[Function]}
      >
        <option
          value="0"
        >
          Choose...
        </option>
        <option
          value="-1"
        >
          High to Low
        </option>
        <option
          value="1"
        >
          Low to High
        </option>
      </select>
    </div>
    <div
      className="hotels"
    >
      <ul
        className="list-unstyled"
      >
        <li
          className="hotel media"
        >
          <img
            alt="Spring Phoenix Resort"
            className="align-self-start mr-3"
            src="https://images.pexels.com/photos/271624/pexels-photo-271624.jpeg?w=1260&h=750&auto=compress&cs=tinysrgb"
          />
          <div
            className="media-body"
          >
            <h1
              className="mt-0"
            >
              Spring Phoenix Resort
            </h1>
            <span>
              5
              <i
                className="fas fa-star"
              />
            </span>
            <p>
              <span>
                car park
                 
              </span>
              <span>
                restaurant
                 
              </span>
              <span>
                bar
                 
              </span>
            </p>
          </div>
        </li>
        <li
          className="hotel media"
        >
          <img
            alt="Triple Pass Hotel"
            className="align-self-start mr-3"
            src="https://images.pexels.com/photos/261102/pexels-photo-261102.jpeg?w=1260&h=750&auto=compress&cs=tinysrgb"
          />
          <div
            className="media-body"
          >
            <h1
              className="mt-0"
            >
              Triple Pass Hotel
            </h1>
            <span>
              3
              <i
                className="fas fa-star"
              />
            </span>
            <p>
              <span>
                car park
                 
              </span>
              <span>
                gym
                 
              </span>
            </p>
          </div>
        </li>
        <li
          className="hotel media"
        >
          <img
            alt="Southern Shack Hotel & Spa"
            className="align-self-start mr-3"
            src="https://images.pexels.com/photos/338504/pexels-photo-338504.jpeg?w=1260&h=750&auto=compress&cs=tinysrgb"
          />
          <div
            className="media-body"
          >
            <h1
              className="mt-0"
            >
              Southern Shack Hotel & Spa
            </h1>
            <span>
              4
              <i
                className="fas fa-star"
              />
            </span>
            <p>
              <span>
                pool
                 
              </span>
              <span>
                bar
                 
              </span>
              <span>
                restaurant
                 
              </span>
            </p>
          </div>
        </li>
        <li
          className="hotel media"
        >
          <img
            alt="Oriental Cove Resort"
            className="align-self-start mr-3"
            src="https://images.pexels.com/photos/6534/holiday-vacation-hotel-luxury.jpg?w=1260&h=750&auto=compress&cs=tinysrgb"
          />
          <div
            className="media-body"
          >
            <h1
              className="mt-0"
            >
              Oriental Cove Resort
            </h1>
            <span>
              2
              <i
                className="fas fa-star"
              />
            </span>
            <p>
              <span>
                car park
                 
              </span>
              <span>
                pool
                 
              </span>
            </p>
          </div>
        </li>
        <li
          className="hotel media"
        >
          <img
            alt="Lunar Hotel"
            className="align-self-start mr-3"
            src="https://images.pexels.com/photos/53577/hotel-architectural-tourism-travel-53577.jpeg?w=1260&h=750&auto=compress&cs=tinysrgb"
          />
          <div
            className="media-body"
          >
            <h1
              className="mt-0"
            >
              Lunar Hotel
            </h1>
            <span>
              4
              <i
                className="fas fa-star"
              />
            </span>
            <p>
              <span>
                pool
                 
              </span>
              <span>
                bar
                 
              </span>
              <span>
                restaurant
                 
              </span>
            </p>
          </div>
        </li>
        <li
          className="hotel media"
        >
          <img
            alt="Rosewood Hotel"
            className="align-self-start mr-3"
            src="https://images.pexels.com/photos/186181/pexels-photo-186181.jpeg?w=1260&h=750&auto=compress&cs=tinysrgb"
          />
          <div
            className="media-body"
          >
            <h1
              className="mt-0"
            >
              Rosewood Hotel
            </h1>
            <span>
              1
              <i
                className="fas fa-star"
              />
            </span>
            <p />
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
